### PR TITLE
Improve concat performance

### LINKED ISSUE #7833: Slow performance of concat()
### What is your issue?

In attempting to concatenate many datasets along a large dimension (total size ~100,000,000) I'm finding very slow performance, e.g., tens of seconds just to concatenate two datasets.

With some profiling, I find all the time is being spend in this list comprehension:

https://github.com/pydata/xarray/blob/51554f2638bc9e4a527492136fe6f54584ffa75d/xarray/core/concat.py#L584

I don't know exactly what's going on here, but it doesn't look right - e.g., if the size of the dimension to be concatenated is large, this list comprehension can run millions of loops, which doesn't seem related to the intended behaviour.

Sorry I don't have an MRE for this yet but please let me know if I can help further.

### PR DESCRIPTION
**Pull request by Illviljan**
* Don't use python for loops for possibly large coords. Rather create a np array once then filter out bad data.
* DuckArrayModule is slightly slow, so cache the first import in a dict instead to speed up later calls.
* Add more typing to be more confident that inputs are valid and then remove redundant checks and conversions.

- [x] Requires #7843 
- [x] Requires #7844.
- [x] Requires #7858
- [x] Closes #7833
- [x] Tests added
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

### CODE REVIEW

File: xarray/core/concat.py [hunk 1]
```diff
@@ -543,7 +544,8 @@ def ensure_common_dims(vars, concat_dim_lengths):
         # ensure each variable with the given name shares the same
         # dimensions and the same shape for all of them except along the
         # concat dimension
-        common_dims = tuple(pd.unique([d for v in vars for d in v.dims]))
+        # common_dims = tuple(pd.unique([d for v in vars for d in v.dims]))
```
Comment (Illviljan) [inline xarray/core/concat.py:547]:
```suggestion
```

**Comment by Illviljan** [comment]
Think I'll stop here. 

@alimanfoo, feel free to try this out if you have the time.

File: xarray/core/dataset.py [hunk 2]
```diff
@@ -6403,7 +6406,14 @@ def to_dask_dataframe(
         columns.extend(k for k in self.coords if k not in self.dims)
         columns.extend(self.data_vars)
 
+        has_many_dims = len(ordered_dims) > 1
```
Comment (dcherian) [inline xarray/core/dataset.py:6409]: Can we move this to a separate PR please?

File: xarray/core/missing.py [hunk 3]
```diff
@@ -709,28 +710,24 @@ def interp_func(var, x, new_x, method: InterpOptions, kwargs):
         ]
         new_x_arginds = [item for pair in new_x_arginds for item in pair]
 
-        args = (
-            var,
-            range(ndim),
-            *x_arginds,
-            *new_x_arginds,
-        )
+        args = (var, range(ndim), *x_arginds, *new_x_arginds)
```
Comment (dcherian) [inline xarray/core/missing.py:713]: can we make this a separate PR too please?

**Review by dcherian — COMMENTED** [review]
Great improvement! What was the most consequential change?

File: xarray/core/concat.py [hunk 4]
```diff
@@ -568,38 +569,40 @@ def get_indexes(name):
...
             for i, ds in enumerate(datasets):
                 if name in ds.variables:
                     variables.append(ds[name].variable)
-                    # add to variable index, needed for reindexing
-                    var_idx = [
-                        sum(concat_dim_lengths[:i]) + k
-                        for k in range(concat_dim_lengths[i])
-                    ]
-                    variable_index.extend(var_idx)
-                    var_concat_dim_length.append(len(var_idx))
```
Comment (Illviljan) [inline xarray/core/concat.py:589]: This was the major bottleneck.

File: xarray/core/concat.py [hunk 5]
```diff
@@ -543,7 +544,7 @@ def ensure_common_dims(vars, concat_dim_lengths):
         # ensure each variable with the given name shares the same
         # dimensions and the same shape for all of them except along the
         # concat dimension
-        common_dims = tuple(pd.unique([d for v in vars for d in v.dims]))
+        common_dims = tuple(dict.fromkeys(d for v in vars for d in v.dims))
```
[the remark this answers is not in the document]
Comment (Illviljan) [reply xarray/core/concat.py:547]: I find the for loop part much more difficult to parse, dicts can't have duplicate keys at least!

pd.unique was like the 2nd major bottleneck. It is not worth it if it's just for a few dims.

File: xarray/core/missing.py (hunk 3 above)
Comment (Illviljan) [reply xarray/core/missing.py:713]: See #7843

File: xarray/core/concat.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Illviljan) [reply xarray/core/concat.py:547]: That code snippet didn't work for me.

Anyway here's the performance comparison:
```python
import itertools
import numpy as np
import pandas as pd
import xarray as xr


t_size = 8000
t = np.arange(t_size)
var = xr.Variable(dims=("T",), data=np.random.randn(t_size)).chunk()

data_vars = {f"long_name_{v}": ("T", var) for v in range(500)}
vars = tuple(v[-1] for v in data_vars.values())

%timeit tuple(pd.unique([d for v in vars for d in v.dims]))
216 µs ± 3.68 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit tuple(dict.fromkeys(d for v in vars for d in v.dims))
133 µs ± 4.57 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

```

File: xarray/core/dataset.py (hunk 2 above)
Comment (Illviljan) [reply xarray/core/dataset.py:6409]: See,  #7844

File: xarray/core/concat.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (keewis) [reply xarray/core/concat.py:547]: > would something like `pd.unique(itertools.chain(v.dims for v in vars))` be better?

this should be
```python
pd.unique(itertools.chain.from_iterable(v.dims for v in vars))
```
or
```python
unique(itertools.chain.from_iterable(v.dims for v in vars))
```
if we add a function calling `tuple(dict.fromkeys(iterable))` to the utils

File: xarray/core/concat.py (hunk 4 above)
Comment (kmuehlbauer) [reply xarray/core/concat.py:589]: Great performance improvement, @Illviljan! It's really helpful to see how the number of datasets/variables effect the performance and how to reduce that drop. I hope I'll keep attention for these issues in future contributions. We've already talked about benchmarks for large number of datasets/variables back in #7400 (https://github.com/pydata/xarray/pull/7400#issuecomment-1378321206), great to have it now.

But seeing all this, I'm wondering if it would be nice to make this missing variable reindexing behaviour available as opt-in (and not by default). So people could gain even more performance if they know their datasets are well formed without missing variables (especially for very large number of datasets).

File: xarray/core/concat.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dcherian) [reply xarray/core/concat.py:547]: We have `OrderedSet` in `utils.py` which would also be more readable. It uses the same trick of assigning to a dict.

File: xarray/core/concat.py [hunk 6]
```diff
@@ -568,38 +569,40 @@ def get_indexes(name):
...
+    variable_index_mask = np.ones(concat_index_size, dtype=bool)
 
     # stack up each variable and/or index to fill-out the dataset (in order)
     # n.b. this loop preserves variable order, needed for groupby.
+    ndatasets = len(datasets)
     for name in vars_order:
         if name in concat_over and name not in result_indexes:
             variables = []
-            variable_index = []
+            mask = variable_index_mask.copy()
```
Comment (dcherian) [inline xarray/core/concat.py:582]: Is it that much  faster to copy than just reassign every time?

File: xarray/core/concat.py (hunk 4 above)
Comment (dcherian) [reply xarray/core/concat.py:589]: Can you add a comment describing the approach please? It still takes some thinking to interpret whats going on

Comment (kmuehlbauer) [reply xarray/core/concat.py:589]: @dcherian Sure, I can add comments at the different locations.

File: xarray/core/concat.py (hunk 6 above)
Comment (Illviljan) [reply xarray/core/concat.py:582]: One of the faster at least, but now that you asked I found `.fill`:
```python
%timeit np.ones(500, dtype=bool)
3.46 µs ± 17.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

a = np.ones(500, dtype=bool)
%timeit a.copy()
429 ns ± 12.5 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit a[:] = True
450 ns ± 1.56 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit a.fill(False)  # Winner! :)
192 ns ± 2.59 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

File: xarray/core/concat.py (hunk 4 above)
Comment (dcherian) [reply xarray/core/concat.py:589]: Sorry I meant @Illviljan

File: xarray/core/concat.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Illviljan) [reply xarray/core/concat.py:547]: Seems slower though:
```python
%timeit OrderedSet(d for v in vars for d in v.dims)
153 µs ± 3.82 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

%timeit dict.fromkeys(d for v in vars for d in v.dims)
121 µs ± 3.04 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

[the remark this answers is not in the document]
Comment (dcherian) [reply xarray/core/concat.py:547]: But pretty minor difference for a function that allocates a big chunk of memory and assigns to it in series, no :)

[the remark this answers is not in the document]
Comment (keewis) [reply xarray/core/concat.py:547]: we could improve that by making use of `dict.update` and `dict.fromkeys`: https://github.com/pydata/xarray/blob/c8c6a4bea656648f048b899dbe8b16f0d3c365aa/xarray/core/utils.py#L539-L540 would become
```python
        self._d.update(dict.fromkeys(values))
```
which allows us to get rid of the explicit loop

File: xarray/core/variable.py [hunk 7]
```diff
@@ -2096,12 +2098,12 @@ def concat(
...
             for var in variables:
-                if var.dims != first_var.dims:
+                if var.dims != first_var_dims:
                     raise ValueError(
-                        f"Variable has dimensions {list(var.dims)} but first Variable has dimensions {list(first_var.dims)}"
+                        f"Variable has dimensions {list(var.dims)} but first Variable has dimensions {list(first_var_dims)}"
                     )
 
-        return cls(dims, data, attrs, encoding)
+        return cls(dims, data, attrs, encoding, fastpath=True)
```
Comment (dcherian) [inline xarray/core/variable.py:2106]: So the fastpath is OK here because know `data` is a duck array? I'm somewhat confused by the fastpath.

Comment (Illviljan) [reply xarray/core/variable.py:2135]: Yeah, my idea is that since `Variable` is an input to the method we can assume `data` is valid. 

I think we can do this in most places where xarray recreates a `Variable`.

File: xarray/core/concat.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Illviljan) [reply xarray/core/concat.py]: Using OrderedSet now. See also #7857.

File: asv_bench/benchmarks/combine.py [hunk 8]
```diff
@@ -2,8 +2,49 @@
...
 import xarray as xr
 
+from . import requires_dask
 
-class Combine:
+
+class Combine1d:
+    """Benchmark concatenating and merging large datasets"""
+
+    def setup(self):
```
Comment (Illviljan) [inline asv_bench/benchmarks/combine.py:11]:
```suggestion
    def setup(self) -> None:
```

File: asv_bench/benchmarks/combine.py [hunk 9]
```diff
@@ -2,8 +2,49 @@
...
+        """Also has to load and arrange t coordinate"""
+        datasets = [self.dsA0, self.dsA1]
+
+        xr.combine_by_coords(datasets)
+
+
+class Combine1dDask(Combine1d):
+    """Benchmark concatenating and merging large datasets"""
+
+    def setup(self):
```
Comment (Illviljan) [inline asv_bench/benchmarks/combine.py:33]:
```suggestion
    def setup(self) -> None:
```

File: xarray/core/concat.py (hunk 4 above)
Comment (Illviljan) [reply xarray/core/concat.py:589]: I've added some comments and maybe some better variable names. Is it any better?

File: xarray/core/dataset.py (hunk 2 above)
Comment (Illviljan) [reply xarray/core/dataset.py]: Done.

File: doc/whats-new.rst [hunk 10]
```diff
@@ -28,7 +28,8 @@ New Features
 
 Breaking changes
 ~~~~~~~~~~~~~~~~
-
+- Improve concatenation performance (:issue:`7833`, :pull:`7824`).
```
Comment (dcherian) [inline doc/whats-new.rst:31]: Move to performance please

File: xarray/core/concat.py [hunk 11]
```diff
@@ -568,38 +569,43 @@ def get_indexes(name):
...
+    variable_index_mask = np.ones(concat_index_size, dtype=bool)
 
     # stack up each variable and/or index to fill-out the dataset (in order)
     # n.b. this loop preserves variable order, needed for groupby.
+    ndatasets = len(datasets)
     for name in vars_order:
         if name in concat_over and name not in result_indexes:
             variables = []
-            variable_index = []
+            variable_index_mask.fill(True)
```
Comment (dcherian) [inline xarray/core/concat.py:583]: A comment for this line would be helpful

File: xarray/core/indexes.py [hunk 12]
```diff
@@ -1495,11 +1495,11 @@ def filter_indexes_from_coords(
     of coordinate names.
 
     """
-    filtered_indexes: dict[Any, Index] = dict(**indexes)
+    filtered_indexes: dict[Any, Index] = dict(indexes)
 
     index_coord_names: dict[Hashable, set[Hashable]] = defaultdict(set)
     for name, idx in indexes.items():
-        index_coord_names[id(idx)].add(name)
```
Comment (dcherian) [inline xarray/core/indexes.py:1502]: This seems sketchy to me, but I'm also not familiar with this code. Can we revert please?

Comment (Illviljan) [reply xarray/core/indexes.py]: Done.

File: xarray/core/dataset.py [hunk 13]
```diff
@@ -1396,11 +1396,14 @@ def _construct_dataarray(self, name: Hashable) -> DataArray:
 
         needed_dims = set(variable.dims)
 
+        coord_name = self._coord_names
         coords: dict[Hashable, Variable] = {}
         # preserve ordering
         for k in self._variables:
-            if k in self._coord_names and set(self.variables[k].dims) <= needed_dims:
```
[the remark this answers is not in the document]
Comment (Illviljan) [reply xarray/core/dataset.py:1402]: Done.

File: doc/whats-new.rst (hunk 10 above)
Comment (Illviljan) [reply doc/whats-new.rst]: Done.

File: xarray/core/concat.py (hunk 11 above)
Comment (Illviljan) [reply xarray/core/concat.py:585]: Done. Is it any better? It's just to reinitialize since we've been replacing that array in-place with False for the previous `name`

**Review by dcherian — APPROVED** [review]
LGTM. Thanks @Illviljan Great PR!
